README.md
# Highcharts for WeeWX extension #

## Description ##

The *Highcharts for WeeWX* extension provides a JSON data feed to support weekly and yearly observation plots using Highcharts. A sample HTML page and the JavaScript necessary to render the Highcharts plots is also included in the [*Highcharts for WeeWX* repositry](https://github.com/gjr80/weewx-highcharts).

The *Highcharts for WeeWX* extension consists of a skin generating two reports, a number of Search List Extensions (SLE) and supporting JavaScript.

## Pre-Requisites ##

The *Highcharts for WeeWX* extension requires *WeeWX v4.5.0* or later. The display of the *Highcharts for WeeWX* extension data on a web page using the included Javascript files requires the Highcharts *Highstock* charting tool and the *JQuery* JavaScript library.

## Installation ##

The preferred method to install the *Highcharts for WeeWX* extension is to use the WeeWX [*wee\_extension* utility](http://weewx.com/docs/utilities.htm#wee_extension_utility). To install the *Highcharts for WeeWX* extension using the WeeWX *wee\_extension* utility:

**Note:** Symbolic names are used below to refer to some file location on the WeeWX system. These symbolic names allow a common name to be used to refer to a directory that may be different from system to system. The following symbolic names are used below:

*$HTML_ROOT*. The path to the directory where WeeWX generated reports are saved. This directory is normally set in the *[StdReport]* section of *weewx.conf*. Refer to [where to find things](http://weewx.com/docs/usersguide.htm#Where_to_find_things) in the WeeWX [User's Guide](http://weewx.com/docs/usersguide.htm) for further information.

1.  Download the latest *Highcharts for WeeWX* extension from the *Highcharts for WeeWX* [releases page](https://github.com/gjr80/weewx-highcharts/releases) into a directory accessible from the WeeWX machine.

	    $ wget -P /var/tmp https://github.com/gjr80/weewx-highcharts/releases/download/v0.4.0/hfw-0.4.0.tar.gz

2.  Install the *Highcharts for WeeWX* extension downloaded at step 1 using the *wee_extension* utility:

    	$ wee_extension --install=/var/tmp/hfw-0.4.0.tar.gz

    **Note:** Depending on your system/installation the above command may need to be prefixed with *sudo*.
        
    This will result in output similar to the following:

        Request to install '/var/tmp/hfw-0.4.0.tar.gz'
        Extracting from tar archive /var/tmp/hfw-0.4.0.tar.gz
        Saving installer file to /home/weewx/bin/user/installer/Hfw
        Saved configuration dictionary. Backup copy at /home/weewx/weewx.conf.20200923124410
        Finished installing extension '/var/tmp/hfw-0.4.0.tar.gz'

3.  Restart WeeWX:

	    $ sudo /etc/init.d/weewx restart

	or

	    $ sudo service weewx restart
	    
    or
    
        $ sudo systemctl restart weewx

This will result in the *Highcharts for WeeWX* JSON data files being generated during each report generation cycle. A default installation will result in the generated JSON data files being placed in the *$HTML_ROOT/json* directory. The *Highcharts for WeeWX* installation can be further customized (eg units of measure, file locations etc) by referring to the *Highcharts for WeeWX* wiki.

## Support ###

General support issues may be raised in the Google Groups [weewx-user forum](https://groups.google.com/group/weewx-user "Google Groups weewx-user forum"). Specific bugs in the *Highcharts for WeeWX* extension code should be the subject of a new issue raised via the [Issues Page](https://github.com/gjr80/weewx-highcharts/issues "Highcharts for WeeWX extension Issues").
 
## Licensing ##

The *Highcharts for WeeWX* extension is licensed under the [GNU Public License v3](https://github.com/gjr80/weewx-highcharts/blob/master/LICENSE "Highcharts for WeeWX extension License").


changelog
v0.4.0
*   now uses .series tags introduced in WeeWX v4.5.0
*   removed SLE entries made redundant due to use of .series tags
*   now requires WeeWX v4.5.0 or later
v0.3.2
*   bindings for appTemp and maxSolarRad are now specified under skin.conf
    [Extras] using apptemp_binding and insolation_binding options
*   replaced getSqlVectors() calls with xtypes.get_series() calls
v0.3.1
*   fixed bug encountered when there are one or more None values returned in
    speed_vec_vt.value
v0.3.0
*   now supports WeeWX v4.0.0 under python 2 or python 3
*   renamed highchartsSearchX.py to highchartssearchlist.py
*   refactored get_extension_list() code in each SLE class
*   merged highcharts.py code into highchartssearchlist.py
*   update to jquery 3.5.1
v0.2.2
*   Fixed bug in getDaySummaryVectors() that resulted in year wind direction 
    plot always displaying 90 degrees.
*   Minor reformatting of comments in various files
v0.2.1
*   Fixed bug with day/week windrose getSqlVectors call that resulted in
    'IndexError: list index out of range' error on line 962 in
    highchartsSearchX.py.
*   Fixed inconsistent naming of [Extra][[WindRose]] stanza in skin.conf
    (this prevented weewx.conf values overriding those in skin.conf)
*   Fixed incorrect [Extras] settings in extension installer (setup.py)
    (this error was masked by the [Extras][[WindRose]] naming issue above)
v0.2.0
*   Removed hard coding of weeWX-WD bindings for appTemp and insolation. Now
    attempts to obtain bindings for each from weeWX-WD, if weeWX-WD is not
    installed bindings are sought in weewx.conf [StdReport][[Highcharts]].
    If no binding can be found appTemp and insolation data is omitted and not
    plotted.
v0.1.0
*   initial release

web server/wxgraphs2.php
<?php
############################################################################
# A Project of TNET Services, Inc. and Saratoga-Weather.org (Canada/World-ML template set)
############################################################################
#
#   Project:    Sample Included Website Design
#   Module:     sample.php
#   Purpose:    Sample Page
#   Authors:    Kevin W. Reed <kreed@tnet.com>
#               TNET Services, Inc.
#
#   Copyright:  (c) 1992-2007 Copyright TNET Services, Inc.
############################################################################
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
############################################################################
#   This document uses Tab 4 Settings
############################################################################
//Version 1.01 - 9-Aug-2011 - added Meteohub support
//Version 1.02 - 24-Mar-2012 - added WeatherCat support
//Version 1.03 - 03-Jul-2012 - added wview support
require_once("Settings.php");
require_once("common.php");
############################################################################
$useUTF8 = false;   // force this page to convert language files to UTF8 for display
$TITLE= $SITE['organ'] . " - " . langtransstr("Trend Graphs");
$showGizmo = true;  // set to false to exclude the gizmo
include("top.php");
############################################################################
?>
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="https://code.highcharts.com/stock/highstock.js"></script>
    <script src="https://code.highcharts.com/stock/highcharts-more.js"></script>
    <script src="https://code.highcharts.com/stock/modules/exporting.js"></script>
    <script src="../scripts/js/weewxtheme.js"></script>
    <script src="../scripts/js/saratogaplots.js"></script>
    <script type="text/javascript">
        window.onload = function() {
            //Get references to links on the page
            var w = document.getElementById("seven_days");
            var y = document.getElementById("calendar_year");

            weekly();

            // Set code to run to display correct plots when the link is clicked
            // by assigning a function to "onclick"
            w.onclick = function() {
                document.getElementById("plottitle").innerHTML="Last Seven Days";
                document.getElementById("plotdesc").innerHTML="These graphs plot the recorded weather observations at this station over the last seven days. The time period displayed by the graph can be changed by using the Zoom buttons, the navigation scrollbar or by clicking and dragging on the graph itself.";
                weekly();
                return false;
            }
            
            y.onclick = function() {
                document.getElementById("plottitle").innerHTML="Last Calendar Year";
                document.getElementById("plotdesc").innerHTML="These graphs plot aggregated weather observations recorded at this station over the last calendar year. The aggregates used  (maximum, minimum, average or sum) are based on a 24 hour day from midnight to midnight. The time period displayed by the graph can be changed by using the Zoom buttons, the navigation scrollbar or by clicking and dragging on the graph itself. Note that for displays over a long period some further grouping of data into week periods may occur, in such cases the 'tooltip' will indicate the the grouping period used.";
                yearly();
                return false;
            }
        }
    </script>
</head>
<body>
<?php
############################################################################
include("header.php");
############################################################################
include("menubar.php");
############################################################################
$graphImageDir = './';
if(isset($SITE['graphImageDir'])) {$graphImageDir = $SITE['graphImageDir']; }
?>

<div id="main-copy">

    <h1><?php langtrans('Trend Graphs'); ?></h1>
    <br/>
    <?php if($SITE['WXsoftware'] <> '') { // software is defined ?>
        <?php print langtrans('Data generated by')." ".$SITE['WXsoftwareLongName']." "; ?>
        <?php if(isset($swversion)) {echo " (".$swversion.")";} ?>
    <?php } else { // software not defined yet ?>
        <p>&nbsp;</p>
        <p>Weather software not specified in configuration.</p>
    <?php } // end software not defined yet ?>
    <hr id="snippets-1" />

    <?php $WWgraphDescs = array(
        'w' => 'Seven_Days',
        'y' => 'Calendar_Year',
    );
    $WWgraphIdx = 'w';  // default is weekly set
    if(isset($_REQUEST['graph'])) { // check out selection
        $req = strtolower($_REQUEST['graph']);
        foreach ($WWgraphDescs as $i => $v) {
            if ($req == $i) {
                $WWgraphIdx = $i;  // default is 7-day set
                break;
            }
        }
    }
    // generate the links for the available periods to display
    print "<p style=\"text-align: center\">";
    foreach ($WWgraphDescs as $i => $v) {
        print "<a id=\"".strtolower($v)."\" href=\"?graph=".$i."\">".langtransstr(str_replace("_", " ", $v))."</a>&nbsp;&nbsp;";
    }
    print "</p>\n"; ?>
    <h3><span id="plottitle">Last Seven Days</span></h3>
    <p><span id="plotdesc">These graphs plot the recorded weather observations at this station over the last seven days. The time period displayed by the graph can be changed by using the zoom buttons, the navigation scrollbar or by clicking and dragging on the graph itself.</span></p>
    <div id="temperatureplot" style="width:99%; height:435px;"></div>
    <br />
    <div id="windchillplot" style="width:99%; height:435px;"></div>
    <br />
    <div id="humidityplot" style="width:99%; height:435px;"></div>
    <br />
    <div id="barometerplot" style="width:99%; height:435px;"></div>
    <br />
    <div id="windplot" style="width:99%; height:435px;"></div>
    <br />
    <div id="winddirplot" style="width:99%; height:435px;"></div>
    <br />
    <div id="rainplot" style="width:99%; height:435px;"></div>
    <br />
    <?php // figure out to display solar and/or UV based on site settings
    if($SITE['SOLAR'] and $SITE['UV']) { // have both sensors ?>
        <div id="radiationplot" style="width:99%; height:435px;"></div>
        <br />
        <div id="uvplot" style="width:99%; height:435px;"></div>
    <?php } elseif ($SITE['SOLAR']) { ?>
        <div id="radiationplot" style="width:99%; height:435px;"></div>
    <?php } elseif ($SITE['UV']) { ?>
        <div id="uvplot" style="width:99%; height:435px;"></div>
    <?php } // end solar/uv selection ?>

</div><!-- end main-copy -->

<?php
############################################################################
include("footer.php");
############################################################################
# End of Page
############################################################################
?>

skins/Highcharts/scripts/weewxtheme.js
/*****************************************************************************

Theme for Weewx Highcharts plots

Based on Gray theme for Highcharts by Torstein Honsi

These are common plot settings across all plots. Change them by all means but
make sure you know what you are doing. The Highcharts API documentation is
your reference.

*****************************************************************************/

Highcharts.theme = {
    chart: {
        backgroundColor: 'rgb(208, 208, 208)',
        borderWidth: 1,
        borderColor: '#000000',
        borderRadius: 8,
        plotShadow: false,
        plotBorderWidth: 0,
    },

    colors: ["#B44242", "#4242B4", "#42B442", "#DF5353", "#aaeeee", "#ff0066", "#eeaaee",
        "#55BF3B", "#DF5353", "#7798BF", "#aaeeee"],

    labels: {
        style: {
            color: '#CCC'
        }
    },

    legend: {
        borderWidth: 0,
        itemStyle: {
            color: '#555'
        },
        itemHoverStyle: {
            color: '#FFF'
        },
        itemHiddenStyle: {
            color: '#999'
        },
        margin: 5,
        padding: 4,
        symbolPadding: 2
    },

    subtitle: {
        style: {
            color: '#555',
            font: '12px Lucida Grande, Lucida Sans Unicode, Verdana, Arial, Helvetica, sans-serif'
        }
    },

    title: {
        margin: 5,
        style: {
            color: '#555',
            font: '16px Lucida Grande, Lucida Sans Unicode, Verdana, Arial, Helvetica, sans-serif'
        }
    },

    xAxis: {
        labels: {
            style: {
                color: '#555',
                fontWeight: 'bold',
                whiteSpace: 'nowrap'
            }
        },
        minorGridLineWidth: 0,
        minorTickInterval: 'auto',
        minorTickLength: 2,
        minorTickWidth: 1,
        minorTickPosition: 'inside',
        tickColor: '#555',
        title: {
            style: {
                color: '#555',
            }
        }
    },
    yAxis: {
        allowDecimals: false,
        gridLineColor: '#DDD',
        gridLineWidth: 1,
        labels: {
            style: {
                color: '#555',
            }
        },
        minorTickColor: '#555',
        minorTickInterval: 'auto',
        minorTickLength: 2,
        minorTickWidth: 1,
        minorTickPosition: 'inside',
        title: {
            style: {
                color: '#555',
                font: 'bold 12px Lucida Grande, Lucida Sans Unicode, Verdana, Arial, Helvetica, sans-serif'
            }
        }
    },

    tooltip: {
        backgroundColor: 'rgba(255, 255, 204, .8)',
        borderWidth: 0,
        dateTimeLabelFormats: {
            hour: '%e %b %H:%M',
            day: '%e %b',
            week: '%e %b',
            month: '%b %y',
            year: '%b %y',
        },
        style: {
            color: '#555'
        }
    },

    plotOptions: {
        line: {
            dataLabels: {
                color: '#CCC'
            },
            marker: {
                lineColor: '#333'
            }
        },
        column: {
            shadow: false
        },
        spline: {
            marker: {
                lineColor: '#333'
            }
        },
        scatter: {
            marker: {
                lineColor: '#333'
            }
        },
        candlestick: {
            lineColor: 'white'
        }
    },

    toolbar: {
        itemStyle: {
            color: '#CCC'
        }
    },

    navigation: {
        buttonOptions: {
            symbolStroke: '#DDDDDD',
            hoverSymbolStroke: '#FFFFFF',
            theme: {
                fill: {
                    linearGradient: { x1: 0, y1: 0, x2: 0, y2: 1 },
                    stops: [
                        [0.4, '#606060'],
                        [0.6, '#333333']
                    ]
                },
                stroke: '#000000'
            }
        }
    },

    // scroll charts
    rangeSelector: {
        buttonTheme: {
            fill: {
                linearGradient: { x1: 0, y1: 0, x2: 0, y2: 1 },
                stops: [
                    [0.4, '#888'],
                    [0.6, '#555']
                ]
            },
            stroke: '#000000',
            style: {
                color: '#CCC',
                fontWeight: 'bold'
            },
            states: {
                hover: {
                    fill: {
                        linearGradient: { x1: 0, y1: 0, x2: 0, y2: 1 },
                        stops: [
                            [0.4, '#BBB'],
                            [0.6, '#888']
                        ]
                    },
                    stroke: '#000000',
                    style: {
                        color: 'white'
                    }
                },
                select: {
                    fill: {
                        linearGradient: { x1: 0, y1: 0, x2: 0, y2: 1 },
                        stops: [
                            [0.1, '#000'],
                            [0.3, '#333']
                        ]
                    },
                    stroke: '#000000',
                    style: {
                        color: 'yellow'
                    }
                }
            }
        },
        inputBoxBorderColor: '#555',
        inputDateFormat: '%e %b %Y',
        inputEditDateFormat: '%e %b %Y',
        inputStyle: {
            backgroundColor: '#EEE',
            color: '#555'
        },
        labelStyle: {
            color: '#555'
        }
    },

    navigator: {
        handles: {
            backgroundColor: '#666',
            borderColor: '#AAA'
        },
        outlineColor: 'yellow',
        maskFill: 'rgba(16, 16, 16, 0.2)',
        xAxis: {
            dateTimeLabelFormats: {
                day: '%e %b',
                week: '%e %b',
                month: '%b %y',
            },
        },
        yAxis: {
            minorTickWidth: 0,
            tickWidth: 0,
        }
    },

    scrollbar: {
        barBackgroundColor: {
                linearGradient: { x1: 0, y1: 0, x2: 0, y2: 1 },
                stops: [
                    [0.4, '#888'],
                    [0.6, '#555']
                ]
            },
        barBorderColor: '#CCC',
        buttonArrowColor: '#CCC',
        buttonBackgroundColor: {
                linearGradient: { x1: 0, y1: 0, x2: 0, y2: 1 },
                stops: [
                    [0.4, '#888'],
                    [0.6, '#555']
                ]
            },
        buttonBorderColor: '#CCC',
        rifleColor: '#FFF',
        trackBackgroundColor: {
            linearGradient: { x1: 0, y1: 0, x2: 0, y2: 1 },
            stops: [
                [0, '#000'],
                [1, '#333']
            ]
        },
        trackBorderColor: '#666'
    },

    // special colors for some of the demo examples
    legendBackgroundColor: 'rgba(48, 48, 48, 0.8)',
    legendBackgroundColorSolid: 'rgb(70, 70, 70)',
    dataLabelsColor: '#444',
    textColor: '#E0E0E0',
    maskColor: 'rgba(255,255,255,0.3)'
};

// Apply the theme
var highchartsOptions = Highcharts.setOptions(Highcharts.theme);


install.py
"""
This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

                     Installer for Highcharts for WeeWX

 Version: 0.4.0                                         Date: xx xxxxx 2021

 Revision History
    xx xxxxx 2021       v0.4.0
        - changed required WeeWX version number
        - removed unused Label, StringFormat and group overrides
    17 March 2021       v0.3.2
        - version number change only
    16 October 2020
        - version number change only
    20 September 2020   v0.3.0
         - now supports WeeWX 4.0.0 under python 2 or python 3
         - removed highcharts.py from install file list
         - renamed highchartsSearchX.py to highchartssearchlist.py
    4 September 2018    v0.2.2
        - version number change only
    16 May 2017         v0.2.1
        - fixed errors in various 'Extras' settings (error previously hidden
          due to [[Windrose]]/[[WindRose]] issue)
    4 May 2017          v0.2.0
        - version number change only
    21 November 2016    v0.1.0
        - initial implementation
"""

import weewx

from distutils.version import StrictVersion
from setup import ExtensionInstaller

REQUIRED_VERSION = "4.5.0a1"
HFW_VERSION = "0.4.0"


def loader():
    return HfwInstaller()


class HfwInstaller(ExtensionInstaller):
    def __init__(self):
        if StrictVersion(weewx.__version__) < StrictVersion(REQUIRED_VERSION):
            msg = "%s requires WeeWX %s or greater, found %s" % ('Hfw ' + HFW_VERSION,
                                                                 REQUIRED_VERSION,
                                                                 weewx.__version__)
            raise weewx.UnsupportedFeature(msg)
        super(HfwInstaller, self).__init__(
            version=HFW_VERSION,
            name='Hfw',
            description='WeeWX support for plotting observational data using Highcharts.',
            author="Gary Roderick",
            author_email="gjroderick<@>gmail.com",
            config={
                'StdReport': {
                    'Highcharts': {
                        'skin': 'Highcharts',
                        'CheetahGenerator': {
                            'ToDate': {
                                'YearJSON': {
                                    'stale_age': '3600'
                                }
                            }
                        },
                        'Units': {
                            'Groups': {
                                'group_altitude': 'meter',
                                'group_degree_day': 'degree_C_day',
                                'group_pressure': 'hPa',
                                'group_rain': 'mm',
                                'group_speed': 'km_per_hour',
                                'group_temperature': 'degree_C'
                            },
                        },
                        'Extras': {
                            'MinRange': {
                                'outTemp': [10, 'degree_C'],
                                'windchill': [10, 'degree_C'],
                                'barometer': [20, 'hPa'],
                                'windSpeed': '10',
                                'rain': [5, 'mm'],
                                'radiation': '500',
                                'UV': '16'
                            },
                            'WindRose': {
                                'title': 'Wind Rose',
                                'source': 'windSpeed',
                                'period': [86400, 604800, 'month', 'year'],
                                'aggregate_type': '',
                                'aggregate_interval': '',
                                'petals': '16',
                                'petal_colors': ['aqua', '0x0099FF', '0x0033FF', '0x009900', '0x00CC00', '0x33FF33', '0xCCFF00'],
                                'speedfactor': ['0.0', '0.1', '0.2', '0.3', '0.5', '0.7', '1.0'],
                                'show_legend_title': 'True',
                                'show_band_percent': 'True',
                                'bullseye_percent': 'True',
                                'precision': '1',
                                'bullseye_size': '20',
                                'bullseye_color': '0xFFFACD',
                                'calm_limit': '0.5'
                            }
                        }
                    }
                }
            },
            files=[('bin/user', ['bin/user/highchartssearchlist.py']
                    ),
                   ('skins/Highcharts', ['skins/Highcharts/json/week.json.tmpl',
                                         'skins/Highcharts/json/year.json.tmpl',
                                         'skins/Highcharts/skin.conf'
                                         ]
                    )
                   ]
        )
